### DISCUSSION TITLE
Add the ability to change the default relationship loading strategy for all models within a `DeclarativeBase`

### DISCUSSION THREAD

**Discussion by caffeinatedMike** [sqlalchemy/sqlalchemy#10853] [Ideas]
I'd like to be able to change the default relationship loading strategy from `lazy="select"` to `lazy="raise"`. I see this as particularly beneficial when utilizing an async database setup because it would enforce explicitly using `.options(selectinload(...))` or `.options(joinedload(...))` to keep queries self-contained and easier to debug.

Ideally this would be able to be configured on the declarative base class or in it's metadata in order to keep the project-wide configuration all in one place.

If this is currently possible somehow Id love to know how (I've tried searching and couldn't find anything on the topic). If it's not currently possible, would this be a feature request worth exploring?

It seems this was [asked about 7 years ago](https://stackoverflow.com/questions/39319398/in-sqlalchemy-how-can-i-configure-all-relationships-to-use-a-loading-strategy) and there still isn't a native solution.

**Comment by caffeinatedMike** [comment]
Well, I feel dumb. ChatGPT to the rescue (yay for AI!). Here's the conversation and solution for anyone else that stumbles upon this issue.

> **Me:**
> Is there a way to change the default loading strategy of sqlalchemy declarative base relationships?
> 
> **ChatGPT:**
> Yes, you can change the default loading strategy of SQLAlchemy declarative base relationships by specifying the lazy parameter when defining the relationship. Options for the lazy parameter include 'select', 'joined', 'subquery', 'dynamic', and others. For example:
> ```python
> class Parent(Base):
>     __tablename__ = 'parents'
>     id = Column(Integer, primary_key=True)
>     children = relationship('Child', lazy='joined')  # Change 'joined' to your desired strategy
> ```
> Choose the loading strategy that best suits your application's performance requirements.
> 
> **Me:**
> No, how would I change it within the DeclarativeBase so that ALL relationships default to a different loading strategy?
> 
> **ChatGPT:**
> To change the default loading strategy for all relationships within a DeclarativeBase, you can customize the __init__ method of your base class. Here's an example:
> ```python
> from sqlalchemy.ext.declarative import declarative_base, DeclarativeMeta
> 
> class CustomBase(declarative_base()):
>     def __init__(cls, name, bases, d):
>         super().__init__(name, bases, d)
> 
>         # Set the default loading strategy for all relationships
>         for attr, value in d.items():
>             if isinstance(value, DeclarativeMeta) and hasattr(value, '__table__'):
>                 for relationship in value.__table__.metadata.tables.values():
>                     for col in relationship.columns:
>                         if hasattr(col, 'property') and hasattr(col.property, 'strategy'):
>                             col.property.strategy = 'select'  # Change 'select' to your desired strategy
> ```
> Now, you can use CustomBase as your base class for your declarative models, and it will set the default loading strategy for all relationships to the specified one. Adjust the col.property.strategy assignment to the loading strategy you want.
> ```python
> class Parent(CustomBase):
>     __tablename__ = 'parents'
>     id = Column(Integer, primary_key=True)
>     children = relationship('Child')  # Relationships will use the default strategy specified in CustomBase
> ```
> Remember to customize the loading strategy according to your specific requirements.
> 
> **Me:**
> And will changing the strategy on a specific relationship still work with this approach or will it override all relationships no matter what if strategy is present on a particular relationship?
> 
> **ChatGPT:**
> Changing the loading strategy on a specific relationship will still work with this approach. The customization in the CustomBase class is setting the default loading strategy for all relationships that don't have an explicitly defined strategy.
> 
> If you specify a loading strategy for a particular relationship in one of your models, it will override the default strategy set in CustomBase for that specific relationship. The customization only comes into play when a relationship doesn't have an explicitly defined strategy.

**Reply by zzzeek** [reply]
that recipe doesn't seem correct at all.  There's no ".strategy"  attribute on relationship, it's called "strategy_key" and it takes the form of a tuple.    strategy_key is also assigned up front in the constructor of relationship() from the "lazy" parameter, which means overwriting it after the relationship was created would blow away the previous setting.

I'm pretty opposed to ChatGPT for this kind of thing so feel free to ping here if/when the ChatGPT solution is not working, then we can talk about it.

**Reply by caffeinatedMike** [reply]
Man, I knew it was too good to be true! 😫 I didn't have a chance to test it, but I know now it wouldn't work. So, any idea how I could control this then?

**Reply by CaselIT** [reply]
Easiest thing for now it so do this https://github.com/sqlalchemy/sqlalchemy/discussions/10853#discussioncomment-8070744

**Comment by CaselIT** [comment]
I think it may make sense, but I'm not sure how it would be configured.
Maybe something similar to how the type_annotation_map is modified?

What do you think mike?

**Reply by zzzeek** [reply]
it feels weird that something would be changing that because the default for `relationship()` is `lazy="select"`, for that to be modified at a distance as per the API seems wrong.

using something similar to what ChatGPT came up with, I would just use the normal API, using attribute instrumentation events and then checking that the value of "lazy" was not already modified to something else.

the other funny thing is that if you're using asyncio, whether or not you have "lazy='raise'", you still have enforced no-lazy-loading set up and you get almost the same stack trace.  this is just a cosmetic concern as to **what** exception is raised.

the simplest way of course remains, just use a function:

```py
from sqlalchemy.orm import relationship as _relationship

def relationship(*arg, **kw):
    kw.setdefault('lazy', "raise")
    return _relationship(*arg, **kw)
```

that above is the clearest and most straightforward, error free way to go.

**Reply by zzzeek** [reply]
the API would have to be changed so that relationship lazy no longer defaults to "select" - it would default to a symbol like ~~"use_registry"~~ "use_mapper" or something like that (or `None`) and then the ~~registry~~ mapper would need to have the default "lazy" setting.  

then, the mechanics of how `relationship()` initializes has to change since "lazy" will no longer be available when the object is first constructed, it will have to set itself to "registry default" and then update it again when the relationship is brought into the mapping.

**Reply by zzzeek** [reply]
this is also awkward because it seems like this would first and foremost be a `__mapper_args__` argument, and we have a separate issue to allow standard `__mapper_args__` arguments for declarative bases or registries...

**Reply by CaselIT** [reply]
> this is also awkward because it seems like this would first and foremost be a `__mapper_args__` argument, and we have a separate issue to allow standard `__mapper_args__` arguments for declarative bases or registries...

that's why it was strange to have in the base. mapper_args seems the obvious thing where this would go.

Do you think opening a 2.x.x use case issue is worth it?

**Reply by caffeinatedMike** [reply]
I do think this is something worth considering for a `__mapper_args__`. My argument for that is being able to determine on a project basis which approach a company/team wants to take (implicit lazy loading vs explicit loading) and being able to ensure that by specifying the style in one place, the `DeclarativeBase`,  seems like a valid business use-case. It aligns with how you can set eager loading of default values. Though I do understand the two are apples and oranges (default values vs loading related data).

**Comment by caffeinatedMike** [comment]
> the other funny thing is that if you're using asyncio, whether or not you have "lazy='raise'", you still have enforced no-lazy-loading set up and you get almost the same stack trace. this is just a cosmetic concern as to what exception is raised.

Can you speak to this a bit more? So, asyncio sessions automatically default to `lazy="raise"` for relationships? Is this mentioned in the asyncio section of the docs? While my problem becomes a non-issue if this turns out to be the default for asyncio sessions, it still may be worth considering as a mapper arg because explicit is always better than implicit imo.

**Reply by CaselIT** [reply]
> So, asyncio sessions automatically default to lazy="raise" for relationships?

No, they still default to `lazy='select'` but in reality it behaves like [`lazy='raise_on_sql'`](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.lazy) since when using asyncio accessing `some_instance.some_relationship` cannot do an io request (since there is no await there) so a different exception is raised, a "missing greenlet" one, but effectively it's the same from a functionality standpoint, ie no lazy load.

It's documented here https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#asyncio-orm-avoid-lazyloads

**Reply by caffeinatedMike** [reply]
Ahh, thank you for pointing that out to me! I've reviewed that section of the documentation, but it took me a few more times of reading (and re-reading) for me to fully understand. Sorry, async is proving to be a fun, and very nuanced, new concept for me.

**Comment by zzzeek** [comment]
this is added in #10859

**Comment by zzzeek** [comment]
law of twos with #10858
